sdk/python: read utf-8 not as expected, UnicodeDecodeError, unexpected end of data

```
state = JuicefsMachine()
folders_0 = state.init_folders()
files_0 = state.create_file(content=b'\xc2\x80', file_name='a', parent=folders_0, umask=18, user='root')
state.readlink(file=files_0, user='root')
state.readlink(file=files_0, user='root')
state.read(encoding='utf-8', errors='strict', file=files_0, length=1, mode='r', offset=0, user='root', whence=0)
state.teardown()

Traceback (most recent call last):
......
    assert self.equal(result1, result2), red(f'read:\nresult1 is {result1}\nresult2 is {result2}')
AssertionError: read:
result1 is (b'\xc2\x80',)
result2 is UnicodeDecodeError:'utf-8' codec can't decode bytes in position 0--1: unexpected end of data
```

#5882

--- a/.github/scripts/pysdk/pysdk_test.py
+++ b/.github/scripts/pysdk/pysdk_test.py
@@ -456,10 +456,10 @@ class ClientParamsTests(unittest.TestCase):
             meta=meta_url,
             cache_dir="/tmp/jfs_test_cache",
             cache_size="100M",
-            cache_partial_only=True
+            cache_partial_only=False
         )
         
-        size_mb = 50
+        size_mb = 48
         test_data = os.urandom(size_mb * 1024 * 1024) 
         with cache_client.open(self.testfile, 'wb') as f:
             f.write(test_data)

--- a/.github/scripts/pysdk/pysdk_test.py
+++ b/.github/scripts/pysdk/pysdk_test.py
@@ -473,7 +473,7 @@ class ClientParamsTests(unittest.TestCase):
         for root, dirs, files in os.walk(cache_dir):
             for file in files:
                 cache_size += os.path.getsize(os.path.join(root, file))
-        self.assertGreaterEqual(cache_size, size_mb * 1024 * 1024)
+        self.assertGreaterEqual(cache_size, size_mb * 1024 * 1024/2)
 
     def test_io_limits(self):
         limited_client = juicefs.Client(

--- a/sdk/python/juicefs/juicefs/juicefs.py
+++ b/sdk/python/juicefs/juicefs/juicefs.py
@@ -436,8 +436,9 @@ class _File(object):
         self.name = path
         self.flag = flag
         self.length = length
-        self.is_closed = False
-        self.off = self.length if 'a' in mode else 0
+        self.closed = False
+        self.append = 'a' in mode
+        self.off = self.length if self.append else 0
 
     def __fspath__(self):
         return self.name
@@ -451,16 +452,13 @@ class _File(object):
     def seekable(self):
         return True
 
-    def closed(self):
-        return self.is_closed
-
     def fileno(self):
         return self.fd
 
     def isatty(self):
         return False
 
-    def read(self, size):
+    def read(self, size=-1):
         """Read at most size bytes, returned as a byes."""
         self._check_closed()
         if self.flag & MODE_READ == 0:
@@ -489,9 +487,18 @@ class _File(object):
         self.off += len(buf)
         return buf
 
+    def readinto(self, buffer):
+        data = self.read(len(buffer))
+        if not data:
+            return 0
+        buffer[:len(data)] = data
+        return len(data)
+
     def write(self, data):
         """Write the string data to the file."""
         self._check_closed()
+        if isinstance(data, memoryview):
+            data = data.tobytes()
         if not isinstance(data, six.binary_type):
             raise TypeError(f"a bytes-like object is required, not '{type(data).__name__}'")
         if not self.writable():
@@ -499,6 +506,8 @@ class _File(object):
 
         if not data:
             return 0
+        if self.append:
+            self.off = self.length
         n = self.lib.jfs_pwrite(c_int64(_tid()), c_int32(self.fd), data, c_int32(len(data)), c_int64(self.off))
         self.off += n
         if self.off > self.length:
@@ -537,36 +546,104 @@ class _File(object):
         self.length = size
         return size
 
+    def flush(self):
+        return
+
     def fsync(self):
         self.lib.jfs_fsync(c_int64(_tid()), c_int32(self.fd))
 
     def close(self):
-        if self.is_closed:
+        if self.closed:
             return
         self.lib.jfs_close(c_int64(_tid()), c_int32(self.fd))
-        self.is_closed = True
+        self.closed = True
 
     def __del__(self):
         self.close()
 
     def _check_closed(self):
-        if self.is_closed:
+        if self.closed:
             raise ValueError('I/O operation on closed file.')
 
+    def readline(self): # TODO: add parameter `size=-1`
+        """Read until newline or EOF."""
+        ls = self.readlines(1)
+        if ls:
+            return ls[0]
+        return b''
+
+    def xreadlines(self):
+        return self
+
+    def readlines(self, hint=-1):
+        """Return a list of lines from the stream."""
+        self._check_closed()
+        if hint == -1:
+            data = self.read(-1)
+        else:
+            rs = []
+            while hint > 0:
+                r = self.read(1)
+                if not r:
+                    break
+                rs.append(r)
+                if r[0] == b'\n':
+                    hint -= 1
+            data = b''.join(rs)
+        return data.splitlines(True)
+
+    def writelines(self, lines):
+        """Write a list of lines to the file."""
+        self._check_closed()
+        self.write(b''.join(lines))
+        self.flush()
 
 class File(object):
     """A JuiceFS file."""
     def __init__(self, lib, fd, path, mode, flag, length, buffering, encoding=None, errors=None):
         self._file = _File(lib, fd, path, mode, flag, length)
-        self.io = io.BufferedReader(self._file, buffer_size=buffering)
+
+        if buffering < 0:
+            buffering = 128<<10
+
+        if buffering == 0:
+            self.raw_io = self._file
+        elif self._file.readable():
+            if self._file.writable():
+                self.raw_io = io.BufferedRandom(self._file, buffer_size=buffering)
+            else:
+                self.raw_io = io.BufferedReader(self._file, buffer_size=buffering)
+        else:
+            self.raw_io = io.BufferedWriter(self._file, buffer_size=buffering)
+
         if encoding:
-            self.io = io.TextIOWrapper(self.io, encoding=encoding, errors=errors)
+            self.io = io.TextIOWrapper(self.raw_io, encoding=encoding, errors=errors)
+        else:
+            self.io = self.raw_io
 
     def __getattr__(self, name):
         return getattr(self.io, name)
 
     def __fspath__(self):
         return self._file.name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+    def next(self):
+        lines = self.readlines(1)
+        if lines:
+            return lines[0]
+        raise StopIteration
 
     def fileno(self):
         return self._file.fd
@@ -607,6 +684,9 @@ def test():
     with v.open("/d/file", "w") as f:
         f.write("hello")
     with v.open("/d/file", 'rb', 5) as f:
+        data = f.readlines()
+        assert data == [b"hello"]
+    with v.open("/d/file", 'rb', 0) as f:
         data = f.readlines()
         assert data == [b"hello"]
     print(list(v.open("/d/file")))

--- a/sdk/python/juicefs/juicefs/juicefs.py
+++ b/sdk/python/juicefs/juicefs/juicefs.py
@@ -72,6 +72,7 @@ def unpack(fmt, buf):
         fmt = "!" + fmt
     return struct.unpack(fmt, buf[: struct.calcsize(fmt)])
 
+
 class JuiceFSLib(object):
     def __init__(self):
         self.lib = cdll.LoadLibrary(os.path.join(os.path.dirname(__file__), "libjfs.so"))
@@ -426,50 +427,20 @@ class Client(object):
         self.lib.free(buf)
         return res
 
+
 class _File(object):
     """A JuiceFS file."""
-    def __init__(self, lib, fd, path, mode, flag, length, buffering, errors):
+    def __init__(self, lib, fd, path, mode, flag, length):
         self.lib = lib
         self.fd = fd
         self.name = path
-        self.append = 'a' in mode
         self.flag = flag
         self.length = length
-        self.errors = errors
-        self.newlines = None
-        self.closed = False
-        self._buffering = buffering
-        if self._buffering < 0:
-            self._buffering = 128 << 10
-        if flag == MODE_READ | MODE_WRITE:
-            self._buffering = 0
-        self._readbuf = None
-        self._readbuf_off = 0
-        self._writebuf = []
-        self.off = 0
-        if self.append:
-            self.off = self.length
+        self.is_closed = False
+        self.off = self.length if 'a' in mode else 0
 
     def __fspath__(self):
         return self.name
-
-    def __enter__(self):
-        return self
-
-    def __exit__(self, exc_type, exc_value, traceback):
-        self.close()
-
-    def __iter__(self):
-        return self
-
-    def __next__(self):
-        return self.next()
-
-    def next(self):
-        lines = self.readlines(1)
-        if lines:
-            return lines[0]
-        raise StopIteration
 
     def readable(self):
         return self.flag & MODE_READ != 0
@@ -481,7 +452,7 @@ class _File(object):
         return True
 
     def closed(self):
-        return self.closed
+        return self.is_closed
 
     def fileno(self):
         return self.fd
@@ -489,52 +460,28 @@ class _File(object):
     def isatty(self):
         return False
 
-    def _read(self, size):
+    def read(self, size):
+        """Read at most size bytes, returned as a byes."""
         self._check_closed()
         if self.flag & MODE_READ == 0:
             raise io.UnsupportedOperation('not readable')
-        # fill buffer
-        if (not self._readbuf or self._readbuf_off == len(self._readbuf)) and size < self._buffering:
-            if not self._readbuf or len(self._readbuf) < self._buffering:
-                self._readbuf = bytes(self._buffering)
-            n = self.lib.jfs_pread(c_int64(_tid()), c_int32(self.fd), self._readbuf, c_int32(self._buffering), c_int64(self.off))
-            if n < self._buffering:
-                self._readbuf = self._readbuf[:n]
-            self._readbuf_off = 0
-        # read from buffer
+        # read directly
         rs = []
         got = 0
-        if self._readbuf and self._readbuf_off < len(self._readbuf):
-            n = len(self._readbuf) - self._readbuf_off
-            if size >= 0 and size < n:
+        while size > 0 or size < 0:
+            n = 4 << 20
+            if size > 0 and size < n:
                 n = size
-            rs.append(self._readbuf[self._readbuf_off:self._readbuf_off+n])
-            self._readbuf_off += n
+            buf = bytes(n)
+            n = self.lib.jfs_pread(c_int64(_tid()), c_int32(self.fd), buf, c_int32(n), c_int64(self.off+got))
+            if n == 0:
+                break
+            if n < len(buf):
+                buf = buf[:n]
+            rs.append(buf)
             got += n
-            size -= n
-        # read directly
-        if size > 0:
-            while size > 0:
-                n = min(size, 4 << 20)
-                buf = bytes(n)
-                n = self.lib.jfs_pread(c_int64(_tid()), c_int32(self.fd), buf, c_int32(n), c_int64(self.off+got))
-                if n == 0:
-                    break
-                if n < len(buf):
-                    buf = buf[:n]
-                rs.append(buf)
-                got += n
+            if size > 0:
                 size -= n
-        elif size < 0:
-            while True:
-                buf = bytes(128 << 10)
-                n = self.lib.jfs_pread(c_int64(_tid()), c_int32(self.fd), buf, c_int32(len(buf)), c_int64(self.off+got))
-                if n == 0:
-                    break
-                if n < len(buf):
-                    buf = buf[:n]
-                rs.append(buf)
-                got += n
         if len(rs) == 1:
             buf = rs[0]
         else:
@@ -542,37 +489,18 @@ class _File(object):
         self.off += len(buf)
         return buf
 
-    def read(self, size=-1):
-        """Read at most size bytes, returned as a string."""
-        return self._read(size)
-
     def write(self, data):
         """Write the string data to the file."""
         self._check_closed()
-        # TODO: buffer for small write
         if not isinstance(data, six.binary_type):
             raise TypeError(f"a bytes-like object is required, not '{type(data).__name__}'")
-        if self.flag & MODE_WRITE == 0:
+        if not self.writable():
             raise io.UnsupportedOperation('not writable')
 
         if not data:
             return 0
-        n = len(data)
-
-        if self.append:
-            self.off = self.length
-        total = len(data)
-        for b in self._writebuf:
-            total += len(b)
-        if total >= self._buffering:
-            self.flush()
-            if len(data) < self._buffering:
-                self._writebuf.append(data)
-            else:
-                self.lib.jfs_pwrite(c_int64(_tid()), c_int32(self.fd), data, c_int32(len(data)), c_int64(self.off))
-        else:
-            self._writebuf.append(data)
-        self.off += len(data)
+        n = self.lib.jfs_pwrite(c_int64(_tid()), c_int32(self.fd), data, c_int32(len(data)), c_int64(self.off))
+        self.off += n
         if self.off > self.length:
             self.length = self.off
         return n
@@ -584,18 +512,12 @@ class _File(object):
         self._check_closed()
         if whence not in (os.SEEK_SET, os.SEEK_CUR, os.SEEK_END):
             raise ValueError(f'invalid whence ({whence}, should be {os.SEEK_SET}, {os.SEEK_CUR} or {os.SEEK_END})')
-        self.flush()
         if whence == os.SEEK_SET:
             self.off = offset
-            self._readbuf = None
         elif whence == os.SEEK_CUR:
             self.off += offset
-            self._readbuf_off += offset
-            if self._readbuf and (self._readbuf_off < 0 or self._readbuf_off >= len(self._readbuf)):
-                self._readbuf = None
         else:
             self.off = self.length + offset
-            self._readbuf = None
         return self.off
 
     def tell(self):
@@ -607,190 +529,60 @@ class _File(object):
         """Truncate the file to at most size bytes.
         Size defaults to the current file position, as returned by tell()."""
         self._check_closed()
-        if self.flag & MODE_WRITE == 0:
+        if not self.writable():
             raise io.UnsupportedOperation('File not open for writing')
-        self.flush()
         if size is None:
             size = self.tell()
         self.lib.jfs_ftruncate(c_int64(_tid()), c_int32(self.fd), c_uint64(size))
         self.length = size
         return size
 
-    def flush(self):
-        """Flush the write buffers of the file if applicable.
-        This does nothing for read-only and non-blocking streams."""
-        if self._writebuf:
-            data = b''.join(self._writebuf)
-            self.lib.jfs_pwrite(c_int64(_tid()), c_int32(self.fd), data, c_int32(len(data)), c_int64(self.off-len(data)))
-            self._writebuf = []
-
     def fsync(self):
-        """Force write file data to the backend storage."""
-        self.flush()
         self.lib.jfs_fsync(c_int64(_tid()), c_int32(self.fd))
-
-    def close(self):
-        """Close the file. A closed file cannot be used for further I/O operations."""
-        if self.closed:
-            return
-        self.flush()
-        self.lib.jfs_close(c_int64(_tid()), c_int32(self.fd))
-        self.closed = True
-
-    def __del__(self):
-        if not self.closed:
-            self.close()
-
-    def _check_closed(self):
-        if self.closed:
-            raise ValueError('I/O operation on closed file.')
-
-    def readline(self): # TODO: add parameter `size=-1`
-        """Read until newline or EOF."""
-        ls = self.readlines(1)
-        if ls:
-            return ls[0]
-        return b''
-
-    def xreadlines(self):
-        return self
-
-    def readlines(self, hint=-1):
-        """Return a list of lines from the stream."""
-        self._check_closed()
-        if hint == -1:
-            data = self._read(-1)
-        else:
-            rs = []
-            while hint > 0:
-                r = self._read(1)
-                if not r:
-                    break
-                rs.append(r)
-                if r[0] == b'\n':
-                    hint -= 1
-            data = b''.join(rs)
-        return data.splitlines(True)
-
-    def writelines(self, lines):
-        """Write a list of lines to the file."""
-        self._check_closed()
-        self.write(b''.join(lines))
-        self.flush()
-
-class File(object):
-    """A JuiceFS file."""
-    def __init__(self, lib, fd, path, mode, flag, length, buffering, encoding=None, errors=None):
-        self._bytes_io = _File(lib, fd, path, mode, flag, length, buffering, errors=errors)
-        self.encoding = encoding
-        self.is_closed = False
-        self.errors = errors
-        self._text_io = io.TextIOWrapper(
-            self._bytes_io,
-            encoding=self.encoding,
-            errors=self.errors
-        )
-
-    def __fspath__(self):
-        return self._bytes_io.name
-
-    def __enter__(self):
-        return self
-
-    def __exit__(self, exc_type, exc_value, traceback):
-        self.close()
-
-    def __iter__(self):
-        return self
-
-    def __next__(self):
-        return self.next()
-
-    def next(self):
-        lines = self.readlines(1)
-        if lines:
-            return lines[0]
-        raise StopIteration
-
-    def fileno(self):
-        return self._bytes_io.fd
-
-    def isatty(self):
-        return False
-
-    def read(self, size=-1):
-        """Read at most size bytes, returned as a string."""
-        if not self._bytes_io.readable():
-            raise io.UnsupportedOperation('not readable')
-        if self.encoding:
-            return self._text_io.read(size)
-        return self._bytes_io.read(size)
-
-    def write(self, data):
-        """Write the string data to the file."""
-        if not self._bytes_io.writable():
-            raise io.UnsupportedOperation('not writable')
-        if self.encoding:
-            if not isinstance(data, str):
-                raise TypeError(f"write() argument must be str, not {type(data).__name__}")
-            return self._text_io.write(data)
-        return self._bytes_io.write(data)
-
-    def readline(self):
-        """Read until newline or EOF."""
-        if self.encoding:
-            return self._text_io.readline()
-        else:
-            return self._bytes_io.readline()
-
-    def readlines(self, hint=-1):
-        """Return a list of lines from the stream."""
-        if self.encoding:
-            return self._text_io.readlines(hint)
-        else:
-            return self._bytes_io.readlines(hint)
-
-    def writelines(self, lines):
-        """Write a list of lines to the file."""
-        if self.encoding:
-            return self._text_io.writelines(lines)
-        else:
-            return self._bytes_io.writelines(lines)
-
-    def seek(self, offset, whence=0):
-        if self.encoding:
-            return self._text_io.seek(offset, whence)
-        return self._bytes_io.seek(offset, whence)
-
-    def tell(self):
-        if self.encoding:
-            return self._text_io.tell()
-        return self._bytes_io.tell()
-
-    def truncate(self, size=None):
-        if self.encoding:
-            return self._text_io.truncate(size)
-        return self._bytes_io.truncate(size)
-
-    def flush(self):
-        if self.encoding:
-            return self._text_io.flush()
-        return self._bytes_io.flush()
-
-    def fsync(self):
-        return self._bytes_io.fsync()
-
-    def __del__(self):
-        if not self.is_closed:
-            self.close()
 
     def close(self):
         if self.is_closed:
             return
+        self.lib.jfs_close(c_int64(_tid()), c_int32(self.fd))
         self.is_closed = True
-        if hasattr(self, '_text_io') and self._text_io is not None:
-            self._text_io.close()
-        self._bytes_io.close()
+
+    def __del__(self):
+        self.close()
+
+    def _check_closed(self):
+        if self.is_closed:
+            raise ValueError('I/O operation on closed file.')
+
+
+class File(object):
+    """A JuiceFS file."""
+    def __init__(self, lib, fd, path, mode, flag, length, buffering, encoding=None, errors=None):
+        self._file = _File(lib, fd, path, mode, flag, length)
+        self.io = io.BufferedReader(self._file, buffer_size=buffering)
+        if encoding:
+            self.io = io.TextIOWrapper(self.io, encoding=encoding, errors=errors)
+
+    def __getattr__(self, name):
+        return getattr(self.io, name)
+
+    def __fspath__(self):
+        return self._file.name
+
+    def fileno(self):
+        return self._file.fd
+
+    def isatty(self):
+        return False
+
+    def fsync(self):
+        """Force write file data to the backend storage."""
+        self.io.flush()
+        return self._file.fsync()
+
+    def close(self):
+        """Close the file. A closed file cannot be used for further I/O operations."""
+        self.io.close()
+        self._file.close()
 
 
 def test():

--- a/sdk/python/juicefs/juicefs/juicefs.py
+++ b/sdk/python/juicefs/juicefs/juicefs.py
@@ -158,7 +158,7 @@ class Client(object):
         fi = FileInfo()
         self.lib.jfs_stat(c_int64(_tid()), c_int64(self.h), _bin(path), byref(fi))
         return os.stat_result((fi.mode, fi.inode, 0, fi.nlink, fi.uid, fi.gid, fi.length, fi.atime, fi.mtime, fi.ctime))
-    
+
     def exists(self, path):
         """Check if a file exists."""
         try:
@@ -267,7 +267,7 @@ class Client(object):
                 infos.append(name)
         self.lib.free(buf)
         return sorted(infos)
-    
+
     def chmod(self, path, mode):
         """Change the mode of a file."""
         self.lib.jfs_chmod(c_int64(_tid()), c_int64(self.h), _bin(path), c_uint16(mode))
@@ -348,11 +348,11 @@ class Client(object):
     def set_quota(self, path, capacity=0, inodes=0, create=False, strict=False):
         """Set the quota of a directory."""
         self._quota(0, path, capacity, inodes, create=create, strict=strict)
-    
+
     def get_quota(self, path):
         """Get the quota of a directory."""
         return self._quota(1, path)
-    
+
     def del_quota(self, path):
         """Delete the quota of a directory."""
         self._quota(2, path)
@@ -426,16 +426,15 @@ class Client(object):
         self.lib.free(buf)
         return res
 
-class File(object):
+class _File(object):
     """A JuiceFS file."""
-    def __init__(self, lib, fd, path, mode, flag, length, buffering, encoding, errors):
+    def __init__(self, lib, fd, path, mode, flag, length, buffering, errors):
         self.lib = lib
         self.fd = fd
         self.name = path
         self.append = 'a' in mode
         self.flag = flag
         self.length = length
-        self.encoding = encoding
         self.errors = errors
         self.newlines = None
         self.closed = False
@@ -471,6 +470,18 @@ class File(object):
         if lines:
             return lines[0]
         raise StopIteration
+
+    def readable(self):
+        return self.flag & MODE_READ != 0
+
+    def writable(self):
+        return self.flag & MODE_WRITE != 0
+
+    def seekable(self):
+        return True
+
+    def closed(self):
+        return self.closed
 
     def fileno(self):
         return self.fd
@@ -533,19 +544,13 @@ class File(object):
 
     def read(self, size=-1):
         """Read at most size bytes, returned as a string."""
-        buf = self._read(size)
-        if self.encoding:
-            return buf.decode(self.encoding, self.errors)
-        else:
-            return buf
+        return self._read(size)
 
     def write(self, data):
         """Write the string data to the file."""
         self._check_closed()
         # TODO: buffer for small write
-        if self.encoding and not isinstance(data, six.text_type):
-            raise TypeError(f'write() argument must be str, not {type(data).__name__}')
-        if not self.encoding and not isinstance(data, six.binary_type):
+        if not isinstance(data, six.binary_type):
             raise TypeError(f"a bytes-like object is required, not '{type(data).__name__}'")
         if self.flag & MODE_WRITE == 0:
             raise io.UnsupportedOperation('not writable')
@@ -553,8 +558,7 @@ class File(object):
         if not data:
             return 0
         n = len(data)
-        if self.encoding:
-            data = data.encode(self.encoding, self.errors)
+
         if self.append:
             self.off = self.length
         total = len(data)
@@ -580,11 +584,6 @@ class File(object):
         self._check_closed()
         if whence not in (os.SEEK_SET, os.SEEK_CUR, os.SEEK_END):
             raise ValueError(f'invalid whence ({whence}, should be {os.SEEK_SET}, {os.SEEK_CUR} or {os.SEEK_END})')
-        if self.encoding:
-            if whence == os.SEEK_CUR and offset != 0:
-                raise io.UnsupportedOperation("can't do nonzero cur-relative seeks")
-            if whence == os.SEEK_END and offset != 0:
-                raise io.UnsupportedOperation("can't do nonzero end-relative seeks")
         self.flush()
         if whence == os.SEEK_SET:
             self.off = offset
@@ -651,7 +650,7 @@ class File(object):
         ls = self.readlines(1)
         if ls:
             return ls[0]
-        return '' if self.encoding else b''
+        return b''
 
     def xreadlines(self):
         return self
@@ -671,15 +670,128 @@ class File(object):
                 if r[0] == b'\n':
                     hint -= 1
             data = b''.join(rs)
-        if self.encoding:
-            return [l.decode(self.encoding, self.errors) for l in data.splitlines(True)]
         return data.splitlines(True)
 
     def writelines(self, lines):
         """Write a list of lines to the file."""
         self._check_closed()
-        self.write(''.join(lines) if self.encoding else b''.join(lines))
+        self.write(b''.join(lines))
         self.flush()
+
+class File(object):
+    """A JuiceFS file."""
+    def __init__(self, lib, fd, path, mode, flag, length, buffering, encoding=None, errors=None):
+        self._bytes_io = _File(lib, fd, path, mode, flag, length, buffering, errors=errors)
+        self.encoding = encoding
+        self.is_closed = False
+        self.errors = errors
+        self._text_io = io.TextIOWrapper(
+            self._bytes_io,
+            encoding=self.encoding,
+            errors=self.errors
+        )
+
+    def __fspath__(self):
+        return self._bytes_io.name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+    def next(self):
+        lines = self.readlines(1)
+        if lines:
+            return lines[0]
+        raise StopIteration
+
+    def fileno(self):
+        return self._bytes_io.fd
+
+    def isatty(self):
+        return False
+
+    def read(self, size=-1):
+        """Read at most size bytes, returned as a string."""
+        if not self._bytes_io.readable():
+            raise io.UnsupportedOperation('not readable')
+        if self.encoding:
+            return self._text_io.read(size)
+        return self._bytes_io.read(size)
+
+    def write(self, data):
+        """Write the string data to the file."""
+        if not self._bytes_io.writable():
+            raise io.UnsupportedOperation('not writable')
+        if self.encoding:
+            if not isinstance(data, str):
+                raise TypeError(f"write() argument must be str, not {type(data).__name__}")
+            return self._text_io.write(data)
+        return self._bytes_io.write(data)
+
+    def readline(self):
+        """Read until newline or EOF."""
+        if self.encoding:
+            return self._text_io.readline()
+        else:
+            return self._bytes_io.readline()
+
+    def readlines(self, hint=-1):
+        """Return a list of lines from the stream."""
+        if self.encoding:
+            return self._text_io.readlines(hint)
+        else:
+            return self._bytes_io.readlines(hint)
+
+    def writelines(self, lines):
+        """Write a list of lines to the file."""
+        if self.encoding:
+            return self._text_io.writelines(lines)
+        else:
+            return self._bytes_io.writelines(lines)
+
+    def seek(self, offset, whence=0):
+        if self.encoding:
+            return self._text_io.seek(offset, whence)
+        return self._bytes_io.seek(offset, whence)
+
+    def tell(self):
+        if self.encoding:
+            return self._text_io.tell()
+        return self._bytes_io.tell()
+
+    def truncate(self, size=None):
+        if self.encoding:
+            return self._text_io.truncate(size)
+        return self._bytes_io.truncate(size)
+
+    def flush(self):
+        if self.encoding:
+            return self._text_io.flush()
+        return self._bytes_io.flush()
+
+    def fsync(self):
+        return self._bytes_io.fsync()
+
+    def __del__(self):
+        if not self.is_closed:
+            self.close()
+
+    def close(self):
+        if self.is_closed:
+            return
+        self.is_closed = True
+        if hasattr(self, '_text_io') and self._text_io is not None:
+            self._text_io.close()
+        self._bytes_io.close()
+
 
 def test():
     volume = os.getenv("JFS_VOLUME", "test")
@@ -753,6 +865,6 @@ def test():
             size += len(t)
     print("read time:", time.time()-start, size>>20)
 
+
 if __name__ == '__main__':
     test()
-

